### PR TITLE
sys/reboot: API for get reboot type used

### PR DESCRIPTION
**Pull request by nvlsianpu**
Added API for fetch reboot type which caused system startup.
This call extract reboot type vale which was used in previous call to the
sys_rebot() call if the platform allow to store the value, otherwise returns the warm reset type.

At last for nRF patform reboot type is stored in the retention register and application/subsytems might make decision on startup deepens on it.

### CODE REVIEW

File: include/zephyr/sys/reboot.h [hunk 1]
```diff
@@ -34,6 +34,22 @@ extern "C" {
...
 extern FUNC_NORETURN void sys_reboot(int type);
 
+/**
+ * @brief Obtain reboot type
+ *
+ * Fetch reboot type which caused system startup.
+ * This call extract reboot type vale which was used in previous call to the
+ * sys_rebot() call, if the platform allow to store the value in
+ * the reset retention storage.
+ * Once value is read, the reboot type in the system is cleared to
```
Comment (nvlsianpu) [inline include/zephyr/sys/reboot.h:44]: Maybe better to introduce call for clear reboot type instead.

File: include/zephyr/sys/reboot.h [hunk 2]
```diff
@@ -34,6 +34,22 @@ extern "C" {
  */
 extern FUNC_NORETURN void sys_reboot(int type);
 
+/**
```
Comment (mbolivar-nordic) [inline include/zephyr/sys/reboot.h:37]: Nothing wrong with this line, but the commit message just says "TO BE DESCRIBED". This should be fixed.

File: lib/os/reboot.c [hunk 3]
```diff
@@ -24,3 +25,13 @@ FUNC_NORETURN void sys_reboot(int type)
 		k_cpu_idle();
 	}
 }
+
+int __weak sys_arch_get_reboot_type(void)
+{
+	return SYS_REBOOT_WARM;
```
Comment (mbolivar-nordic) [inline lib/os/reboot.c:31]: This seems like an incorrect default implementation. I'm not exactly sure what the right thing to do is, though. But perhaps you could add a new SYS_REBOOT_UNKNOWN and define it to -1 or something like that?

File: include/zephyr/sys/reboot.h (hunk 1 above)
Comment (mbolivar-nordic) [reply include/zephyr/sys/reboot.h:44]: > Maybe better to introduce call for clear reboot type instead.

Have you considered this from the user perspective? What would fit best with the use cases you are enabling with this API?

**Comment by nvlsianpu** [comment]
If someone had idea that the `hwinfo` driver abstarction might be used for geting the reboot-type. See my concerns on that  i have after looking into that abstraction:

I looked into the `hwinfo` driver interface - it looks like not relevant for extracting the reason because of which, goal for which the application had decided to reset the SoC or the level to which the system should be rebooted (basically for what the reboot-type can be used).
hwinfo_get_supported_reset_cause() call is essentially for get the method/event or HW entity which caused the system reset. These information are extracted based on dedicated SoC registers content. Moreover reset using sys_reboot() will be recognized always as `RESET_SOFTWARE` by that call.
The reboot-type in context of this patch is the software defined concept. That why I think we should use new API for extracting it.

File: include/zephyr/sys/reboot.h (hunk 1 above)
Comment (nvlsianpu) [reply include/zephyr/sys/reboot.h:44]: I will go with separate call for clear the reboot type.

File: include/zephyr/sys/reboot.h [hunk 4]
```diff
@@ -34,6 +34,22 @@ extern "C" {
...
+ * @brief Obtain reboot type
+ *
+ * Fetch reboot type which caused system startup.
+ * This call extract reboot type vale which was used in previous call to the
+ * sys_rebot() call, if the platform allow to store the value in
```
Comment (carlescufi) [inline include/zephyr/sys/reboot.h:42]: sys_reboot

File: include/zephyr/sys/reboot.h [hunk 5]
```diff
@@ -34,6 +34,22 @@ extern "C" {
...
+ * the reset retention storage.
+ * Once value is read, the reboot type in the system is cleared to
+ * the default value (SYS_REBOOT_WARM).
+ *
+ * Otherwise it returns the default value which is SYS_REBOOT_WARM.
+ *
+ * @retval reboot type
+ */
+int sys_get_reboot_type(void);
```
Comment (carlescufi) [inline include/zephyr/sys/reboot.h:51]: `sys_reboot_get_type()`

**Comment by henrikbrixandersen** [comment]
> I looked into the `hwinfo` driver interface - it looks like not relevant for extracting the reason because of which, goal for which the application had decided to reset the SoC or the level to which the system should be rebooted (basically for what the reboot-type can be used).
> hwinfo_get_supported_reset_cause() call is essentially for get the method/event or HW entity which caused the system reset. These information are extracted based on dedicated SoC registers content. Moreover reset using sys_reboot() will be recognized always as `RESET_SOFTWARE` by that call.

But it doesn't have to? We could extend the `hwinfo` API to allow differentiating warm/cold reboots where supported. To me, there seems to be too much overlap with the `hwinfo` API in the proposed API.

> The reboot-type in context of this patch is the software defined concept. That why I think we should use new API for extracting it.

The various `hwinfo` reset causes are also software defined and mapped by the platform-specific driver.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Too much overlap with existing `hwinfo` API. Blocking until we have further discussion on this.

**Comment by alexanderwachter** [comment]
I tend to agree with @henrikbrixandersen.
The reset cause is a bit map an can easily be extended with a warm/cold bit

**Comment by henrikbrixandersen** [comment]
So the nRF series distinguish between a warm and a cold reboot and store this information to allow retrieval on the next boot. Do we have examples of other implementations or SoCs, which make this distinction?

**Comment by carlocaione** [comment]
> Do we have examples of other implementations or SoCs, which make this distinction?

Well, pretty much all the bigger SoCs (ARM64) have this capability. For example AmLogic (on which I used this feature in some previous job).

**Comment by carlescufi** [comment]
API meeting:

- @nashif prefers to keep this self-contained in the sys/reboot namespace
- @henrikbrixandersen leans toward
- @gregshue suggests keeping everything in the sys/reboot but adding the info from hwinfo in the `sys_get_reboot_type()`
- @gmarull states that the documentation for the two types of reboot (warm and cold) needs to describe what those mean exactly. Leans towards using hwinfo

**Comment by gregshue** [comment]
🤔 It seems that a rebooted system needs to know why it was rebooted and what it needs to accommodate (reinitialize) on startup.

Some of this info must come from outside the previously-running executable (e.g., HW Watchdog tripped, another agent asserted the reset line).  This is currently provided by the `hwinfo_` API, but I have not found API documentation for what `RESET_SOFTWARE`, `RESET_DEBUG`, and `RESET_USER` mean and how they are to be used. These definitions may matter in considering this PR.

Others may come from catastrophic failures in the previously running executable (e.g., memory access failure, unhandled divide by zero, kernel panic). Triaging the latter events require **additional data be carried across a reset** (e.g., FW build info/file/line, reason, context/instance data).  There is no guarantee that the previously running executable is the same binary as now running, so it is very likely an `int` is insufficient storage. There is also no guarantee that the non-reset storage is valid.

It seems we need an API that at least:

- specifies how to configure the maximum sized data to be retained across reset
- includes an in/valid indication
- includes the size of the data
- specifies a data encoding for common, generic reset conditions
- specifies a data encoding mechanism for (downstream) **component-defined data**

The one currently in this PR seems to be an intermediate step at best.

Thoughts?

**Comment by nvlsianpu** [comment]
@henrikbrixandersen Looking at conversation history there I would go with introducing new API for SoC retention register/memory.
Using that I can rework sys_reboot() for nRF and re-implement https://github.com/zephyrproject-rtos/zephyr/pull/46250

**Pull request closed without merging**